doc/kuksa_requirementes.md

```diff
@@ -0,0 +1,787 @@
+  * field -> **`repeated DataEntryError errors`** from **`GetResponse`**
+  * fielf -> **`repeated DataEntryError errors`** from **`SetResponse`**
+
+```protobuf
+service VAL {
+  rpc Get(GetRequest) returns (GetResponse);
+
+  rpc Set(SetRequest) returns (SetResponse);
+
+  rpc Update(stream SetRequest) returns (SetResponse);
+
+  rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse);
+
+  rpc Actuate(ActuateRequest) returns (ActuateResponse);
+
+  rpc GetMetadata(MetadataRequest) returns (MetadataResponse);
+
+  rpc GetServerInfo(GetServerInfoRequest) returns (GetServerInfoResponse);
+}
+
+message EntryRequest {
+  string path                     = 1;
+  View view                       = 2;
+  repeated Field fields           = 3;
+}
+
+message GetRequest {
+  repeated EntryRequest entries   = 1;
+}
+
+message GetResponse {
+  repeated DataEntry entries      = 1;
+  Error error                     = 1;
+}
+
+message EntryUpdate {
+  DataEntry entry                 = 1;
+  repeated Field fields           = 2;
+}
+
+message SetRequest {
+  repeated EntryUpdate updates    = 1;
+}
+
+message SetResponse {
+  Error error                     = 1;
+}
+
+message SubscribeEntry {
+  string path                     = 1;
+  View view                       = 2;
+  repeated Field fields           = 3;
+}
+
+message SubscribeRequest {
+  repeated SubscribeEntry entries = 1;
+}
+
+message SubscribeResponse {
+  repeated EntryUpdate updates    = 1;
+}
+
+message ActuateRequest {
+  EntryUpdate actuator = 1;
+}
+
+message ActuateResponse {
+  Error error                     = 1;
+}
+
+message MetadataRequest {
+  repeated string path            = 1;
+}
+
+message MetadataResponse {
+  repeated DataEntry metadata     = 1;
+}
+
+message GetServerInfoRequest {
+  // Nothing yet
+}
+
+message GetServerInfoResponse {
+  string name                     = 1;
+  string version                  = 2;
+}
+``` 
+
+# Functional requirements
+
+### As Client
+1. As a client I want to have a single service point accessible from my execution environment.
+   
+2. As a client I want to `Get` sensor values or actuator current values of the vehicle and its environment in a uniform manner.
+  - Details:
+    * All sensor and actuator values for which a client have permission to.
+    * Client must provide the path, paths or wildcard.
+    * I want to receive exactly all the signals’ values I requested without errors.
+  
+3. As a client I want to `Subscribe` to sensor values or actuator current values when some signal value changed -> what about continuous subscription?
+  - Details:
+    * All sensor and actuator values for which a client have permission to.
+    * Client must provide the path, paths or wildcard paths.
+    * I want to receive exactly all the signals’ values I requested without errors.
+
+4. As a client I want to `Subscribe` and be able to apply a filter to receive a sensor value with an interval of x milliseconds only if the value has changed.
+   
```

Comment:
Not sure where to place that:
n. A signal can have value out of the set of the defined value restriction/data type but it can also have those additional failure states:
    - "unknown": The signal with the specified path/name is not defined on the system.
    - "access denied": The signal is basically known, but the consumer/provider has no permission to access it.
    - "not provided": The signal is basically known, but currently no provider is serving it or the provider lost the connection to the backend component serving this signal.
    - "invalid": The signal is being served by a provider, but currently has no valid value. This may happen if a specific value cannot be determined because of some (temporary) physical sensor error or a target value is currently not known, e.g. if a window is moved by a passenger via up/down button.

The failure states "unknown" and "access denied" are real error situations, that shall be avoided by a proper system configuration. Nevertheless, they may occur in case of clients requesting "optional signals" and also if users of the system enter some private mode and some signals accessible before may get inaccessible.

The states "not provided" and "invalid" are regular states that may occur during normal operation of the running system and must be properly handled by clients.

Reply:
Do we need or want to be more clear on how this interval requirement should be interpreted and implemented? 
I can see some different implementations that give partially different behavior

- Poll every x millisecond but send only if value is different compared to the last value sent to client
- Whenever databroker receives a changed value (compared to last value in databroker) check if x millisconds has passed, if so send value to client
- Whenever Databroker receives a value and x milliseconds has passed, compare with previous value sent to client and if changed send to client

Do we want alreayd in requirements to state if implementation shall use a poll-based or event-based mechanism, or do we leave that to the implementation?

Reply:
Answering to the comment by @erikbosch 

Is the requirement `with an interval of x miliseconds` a new requirement (feature request)? I was not aware that the current version of the databroker supports this. 

Intuitively, to me subscribe still has the semantic of only getting informed about changed values. Because of that,  I would go for the third option to check what has been sent to the client before and potentially adding "grace period" before after the last value has been sent before sending another value.

But I get that there are uses cases as well where the client is interested in the value at a fixed frequency to essentially downsample the data stream coming from the sensor.